extensions/razer/internal.m

```diff
@@ -0,0 +1,1658 @@
+    
+    lua_pushvalue(L, 1);
+    lua_pushboolean(L, result);
+    return 2;
+}
+
+#pragma mark - Lua<->NSObject Conversion Functions
+
+// NOTE: These must not throw a Lua error to ensure LuaSkin can safely be used from Objective-C delegates and blocks:
+
+static int pushHSRazer(lua_State *L, id obj) {
+    HSRazer *value = obj;
+    value.selfRefCount++;
+    void** valuePtr = lua_newuserdata(L, sizeof(HSRazer *));
+    *valuePtr = (__bridge_retained void *)value;
+    luaL_getmetatable(L, USERDATA_TAG);
+    lua_setmetatable(L, -2);
+    return 1;
+}
+
+static id toHSRazerFromLua(lua_State *L, int idx) {
+    LuaSkin *skin = [LuaSkin sharedWithState:L];
+    HSRazer *value;
+    if (luaL_testudata(L, idx, USERDATA_TAG)) {
+        value = get_objectFromUserdata(__bridge HSRazer, L, idx, USERDATA_TAG);
+    } else {
+        [skin logError:[NSString stringWithFormat:@"expected %s object, found %s", USERDATA_TAG,
+                        lua_typename(L, lua_type(L, idx))]];
+    }
+    return value;
+}
+
+#pragma mark - Hammerspoon/Lua Infrastructure
+
+static int userdata_tostring(lua_State* L) {
+    LuaSkin *skin = [LuaSkin sharedWithState:L];
+    HSRazer *razer = [skin luaObjectAtIndex:1 toClass:"HSRazer"];
+    NSString *name = razer.name;
+    BOOL isConnected = [razer isConnected];
+    NSString *connected = @"Connected";
+    if (!isConnected) {
+        connected = @"Disconnected";
+    }
+    [skin pushNSObject:[NSString stringWithFormat:@"%s: %@ - %@ (%p)", USERDATA_TAG, name, connected, lua_topointer(L, 1)]];
+    return 1;
+}
+
+static int userdata_eq(lua_State* L) {
+    // Can't get here if at least one of us isn't a userdata type, and we only care if both types are ours, so use luaL_testudata before the macro causes a Lua error:
+    if (luaL_testudata(L, 1, USERDATA_TAG) && luaL_testudata(L, 2, USERDATA_TAG)) {
+        LuaSkin *skin = [LuaSkin sharedWithState:L];
+        HSRazer *obj1 = [skin luaObjectAtIndex:1 toClass:"HSRazer"];
+        HSRazer *obj2 = [skin luaObjectAtIndex:2 toClass:"HSRazer"];
+        lua_pushboolean(L, [obj1 isEqualTo:obj2]);
+    } else {
+        lua_pushboolean(L, NO);
+    }
+    return 1;
+}
+
+// User Data Garbage Collection:
+static int userdata_gc(lua_State* L) {
+    HSRazer *obj = get_objectFromUserdata(__bridge_transfer HSRazer, L, 1, USERDATA_TAG);
+    if (obj) {
+        obj.selfRefCount--;
+        if (obj.selfRefCount == 0) {
+            LuaSkin *skin = [LuaSkin sharedWithState:L];
+            // Stop the Event Tap:
+            [obj destroyEventTap];
+            
+            // Stop HID Manager and perform garbage collection:
+            [obj stopHIDManager];
+            [obj doGC];
+            
+            obj.callbackRef = [skin luaUnref:refTable ref:obj.callbackRef];
+
+            // Disconnect Callback:
+            if (obj.callbackToken != nil) {
+                //[obj.razer close];
+                obj.callbackToken = nil;
+            }
+            obj = nil;
```

Comment:
I would think `obj.nil` would want to go after `obj.lsCanary = tmplsCanary;` a few lines below?